### ISSUE TITLE
Javascript enhanced transforms.

### ISSUE THREAD

**Issue by DylanPiercey** [postcss/postcss#577]
I understand that postcss and is purely preprocessing css at the moment. I think that this is extremely limiting, specifically for things like css polyfills. What I am about to propose may be out of scope for this project but I just want to get people thinking about polyfilling the more tricky features of css4/beyond.

I'm thinking of stuff like this:

```css
/* Input */
.myclass {
    position: sticky;
    top: 10px;
    color: green;
}

/* Output */
.__sticky {
    /* Generic sticky styles if needed */
    position: fixed;
}

.myclass {
    color: green;
}
.__sticky.myclass {
/* Selector specific styles */
    top: 10px;
}
```

In this case the generated javascript would look something like this:

```javascript
// We could output our lists of selectors and some config that require js and generate a javascript polyfill during css compilation.

const POST_CSS = {
   "sticky": [
       {
             selector: ".myClass",
             options: {
                  top: 10px;
             }
       }
    ]
}; 

// And then we could have custom javascript handlers for any transform that needs it.
window.addEventListener("scroll", function () {
    POST_CSS.sticky.forEach(function (rule) {
        let matches = document.querySelectorAll(rule.selector);
        for (let el of matches) {
            if (window.pageYOffset > el.offsetTop + rule.options.top) {
                el.addClass(".__sticky");
            } else {
                el.removeClass(".__sticky");
            }
        }
    });
});
```

The idea here is that we can use postcss to generate both css and some related javascript. Now this example wouldn't be terribly efficient (although it could be much better). However I think that this is something that the css community as a whole should start exploring.

There are some libraries out there that attempt to manually parse the css at runtime and provide transforms but this is costly. I think there could be mostly performant polyfills for pretty much any new or custom css feature if we were to start leveraging both preprocessing and javascript.

What do you guys think?

---

This was also posted [here](https://github.com/cssnext/cssnext/issues/221)

**Comment by ben-eb** [comment]
You're right, it's out of scope. However, you can write a module that does what you're looking for in terms of JS output, and then using PostCSS to iterate a stylesheet in order to find the rules that should be transformed by polyfill.

You might like to team up with @cie as he was also looking into polyfilling. :smile:

Further reading: https://github.com/postcss/postcss/issues/476

**Comment by ai** [comment]
I think it is a good vector. Right now PostCSS API can do by sending callback as plugin option:

```js
postcss([
  elementMq({ saveJs: js => fs.writeFileSync('from-css.js', js) })
])
```

**Comment by trusktr** [comment]
I know this thread is old, but it'd be great if a tool such as PostCSS had a standard way for plugins to hook into polyfilling hooks.

For polyfills to work properly, not only do the polyfills need to operate on the top-level document, but also inside shadow DOM `ShadowRoot`s, and possibly in other ways like performing certain types of rendering with canvas.

Maybe a tool could be written on top of postcss that can standardize how to hook into different features:

- plugins could specify what JS needs to run on a root document, then the system automatically handles patching `attachShadow` in order to ensure that the polyfill runs in all necessary places
- a plugin might request a canvas render, and the system would automatically absolutely position a canvas on top of the target element(s) to add visual features.
- In capable browsers, a plugin might request rendering support via worklets, and the system would automatically create those in various places like `background` properties on the element or on absolutely-positioned `::before` and `::after` content.
- some plugins need to write custom attributes or class names to elements in order to emulate CSS4 selectors like `:has()` or `:focus-within`. The plugins could tell the system what they need to place, then the system worries about placing the attributes/classes on elements in a standard way (including within shadow roots).

Basically the system would standardize the various ways to hook into generic browser features in order to achieve polyfills in a standard way.
